### PR TITLE
Don't fix access of an 'open' override in a 'public' extension

### PR DESCRIPTION
**Pull request by jrose-apple**
This is reasonable to diagnose with a warning, but dropping the 'open' down to 'public' isn't the right fix, because now it's not a valid override. The declaration has to get moved to another extension instead, or the extension has to not set a default access level.

This turned out to be a source compat issue because the same logic that emits the fix-it also updates the access of the member, which then resulted in "must be as accessible as the declaration it overrides" in the *same* build. It's not immediately clear what caused this; probably something's just being validated in a different order than it was before. The change makes sense either way.

rdar://problem/47557376
rdar://problem/28493971

### CODE REVIEW

**Comment by jrose-apple** [comment]
@dingobye, you'd also be a good person to review, since you made changes here recently!

@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

File: test/ClangImporter/objc_override.swift [hunk 1]
```diff
@@ -116,6 +116,16 @@ class MyHashableNSObject: NSObject {
...
 }
 
+// rdar://problem/47557376
+// Adding an override to someone else's class in an extension like this isn't
+// really sound, but it's allowed in Objective-C too.
+extension OverrideInExtensionSub {
+  open override func method() {}
+}
+public extension OverrideInExtensionSub {
+  open override func accessWarning() {} // expected-warning {{declaring open instance method in a public extension}}
```
Comment (brentdax) [inline test/ClangImporter/objc_override.swift]: I know this diagnostic already exists, but "declaring open instance method in a public extension" isn't great. It doesn't really give you any idea of what the compiler will do with this code or how you should correct the issue. Perhaps the fix-it normally provides enough additional information to understand it, but there's no fix-it here. I think it should say something like "public keyword on extension will be ignored for this open instance method".

File: test/ClangImporter/objc_override.swift [hunk 2]
```diff
@@ -116,6 +116,16 @@ class MyHashableNSObject: NSObject {
...
 }
 
+// rdar://problem/47557376
+// Adding an override to someone else's class in an extension like this isn't
+// really sound, but it's allowed in Objective-C too.
+extension OverrideInExtensionSub {
+  open override func method() {}
+}
+public extension OverrideInExtensionSub {
+  open override func accessWarning() {} // expected-warning {{'open' modifier conflicts with extension's default access of 'public'}}
```
Comment (jrose-apple) [inline test/ClangImporter/objc_override.swift]: @brentdax How does this sound? It's still vague about how it's going to be treated for now, but at least it sounds like a problem.

Comment (brentdax) [reply test/ClangImporter/objc_override.swift]: Much better!

**Comment by jrose-apple** [comment]
[waiting for the previous tests to finish before starting more tests, since the only thing that changed was the diagnostic message]

File: lib/Sema/TypeCheckAttr.cpp [hunk 3]
```diff
@@ -1640,10 +1640,15 @@ void AttributeChecker::visitAccessControlAttr(AccessControlAttr *attr) {
...
                                 attr->getAccess(),
-                                D->getDescriptiveKind(),
                                 extAttr->getAccess());
-        swift::fixItAccess(diag, cast<ValueDecl>(D), defaultAccess, false,
-                           true);
+        // Don't try to fix this if the declaration in question is an override
+        // (which can happen with Objective-C extensions). That can lead to a
+        // diagnostic fight between this and "declaration must be as accessible
+        // as the declaration it overrides".
+        if (!D->getAttrs().hasAttribute<OverrideAttr>()) {
```
[the remark this answers is not in the document]
Comment (jrose-apple) [reply lib/Sema/TypeCheckAttr.cpp]: Hrm. Maybe it'd be better to just remove the fix-it altogether. It's hard to tell whether something's trying to satisfy a requirement.

[the remark this answers is not in the document]
Comment (dingobye) [reply lib/Sema/TypeCheckAttr.cpp]: Agree! It is always better to be quiet than misleading.

[the remark this answers is not in the document]
Comment (jrose-apple) [reply lib/Sema/TypeCheckAttr.cpp]: We *will* still have the fighting cases for now in the error cases, but, well…that's not a regression.

[the remark this answers is not in the document]
Comment (dingobye) [reply lib/Sema/TypeCheckAttr.cpp]: Well, that is an orthogonal issue. 'type_witness_not_accessible_proto' keeps warning even though foo() is already declared `public`.

**Review by dingobye — APPROVED** [review]
LGTM!

**Pull request merged**